### ISSUE TITLE
When to use a console message, and when to use a dialog box

### ISSUE THREAD

**Issue by jordanbrown0** [openscad/openscad#4631]
In #4600 a user is confused because "export to STL" wasn't doing anything.  It turned out that the problem was that the user hadn't done a render, and so **on the console** OpenSCAD was displaying "ERROR: Nothing to export! Try rendering first (press F6)".  The user simply hadn't noticed this message on the console.

In my opinion, this is a UI design error.  Having **execution** warnings and errors appear in the console seems natural, but when you do a UI action - click a button, select a menu item, use a keystroke shortcut - if there is an error then it seems like the result should be a dialog box that either tells you about the problem and you then dismiss, or tells you about the problem and guides you through resolving it.

In the case in #4600, it seems to me that the minimum should have been a dialog box with that message in it, but better still would have been for that dialog box to have a RENDER button that would render the model and then export it.

**Comment by gsohler** [comment]
Of Course I admit that it would be better to have this warning message not in the console, but having something which pops into your eyes.
But after all, OpenSCAD is a programming language: You insert your code, you watch your model and  in case it does not look expected you watch into the console more more details.
Personally I don't bear with the user which did not read the message but I agree that it would be better to make a programmer's tool more windows like ;)

**Comment by jordanbrown0** [comment]
The distinction is between messages that result *from your program* and messages that result *from your actions in the UI*.  Messages that result from your program belong in the console; messages that result from UI actions belong in the UI, typically as popups.

**Comment by MethylBromide** [comment]
I'm with Jordan on this.

**Comment by MethylBromide** [comment]
And in addition: we lack a good way to respond to illegal inputs in the Customizer. This too is the result of a user action rather than a code issue, and the choices on how to respond in the script are:

- Log a warning using echo, which the user might not see.
- Log a warning using assert, which the user might be a little more likely to see but which unnecessarily adds a stack trace.
- Create some "ghost" text in the preview window, which the user is more likely to notice but which is a challenge to position so that it doesn't intersect any objects in the coordinate space.

**Comment by bertiebaggio** [comment]
> In https://github.com/openscad/openscad/issues/4600 a user is confused because "export to STL" wasn't doing anything. It turned out that the problem was that the user hadn't done a render, and so on the console OpenSCAD was displaying "ERROR: Nothing to export! Try rendering first (press F6)". The user simply hadn't noticed this message on the console.

I ran into this today in my first interaction with OpenSCAD (nightly build OpenSCAD-2026.04.26-x86_64.AppImage on Linux), which was otherwise excellent. The error message in the console is helpful, but being new I didn't know to look there; in fact I was looking at the controlling terminal from which I had launched the appimage for other reasons! †

I agree with Jordan Brown's reasoning in terms of context: it makes sense to me as a user to have UI-originating messages be displayed by the UI. Also displaying the messages in the console would be fine, of course. 

If, for the sake of discussion an alternative or additional option is desired, a small UI 'export status' widget would also be a reasonable option; which could report, say:

 - [green] ✓ File successfully exported to /path/to/example_filename.sty
 - [red] ✗ File not exported: $reason 

Cheers :)

(†: I was having a different UI issue relating to fractional `QT_SCALE_FACTOR` which I am happy to report separately)

**Comment by MichaelAtOz** [comment]
> an alternative or additional option is desired, a small UI 'export status' widget

If the user doesn't notice the console are they going to notice a small* widget? 
*I'd say a widget should not occupy meaningful screen real estate, at biggest size of editor icons.

Is it possible to flash the console (invert fg/bg colour briefly) if an error/warning occured in the last operation??

**Comment by jordanbrown0** [comment]
Yeah, I wouldn't be too optimistic about noticing and correctly interpreting an icon.

Yes, it's certainly possible to flash the console, but what if the console isn't visible at all?  A popup seems like a much more usual UI idiom, doesn't take up any screen space at all in the normal case, and can have actual words explaining what the problem is.

**Comment by bertiebaggio** [comment]
> If the user doesn't notice the console are they going to notice a small* widget?

Fair question. On its own a small icon /widget would probably not be noticed, though a distinct visual change might draw attention. 

For the sake of comparison, as you may know Blender uses a small floating widget, though I believe they use their own custom UI toolkit.

<img width="421" height="88" alt="Image" src="https://github.com/user-attachments/assets/a761b24f-8ba3-47c0-810c-960a7bcb8960" />

https://github.com/user-attachments/assets/896845b6-860c-4a7f-b9a0-2bb62f3c8401

While distinct from the surrounding UI, it is quite small and could easily be missed by someone unfamiliar with the program.

In the case of an error, the message is written both to the status widget and appears as a message at the cursor position:

<img width="2193" height="732" alt="Image" src="https://github.com/user-attachments/assets/08fef3f9-051b-45a8-b44b-07f16ab0e849" />

----

>  A popup seems like a much more usual UI idiom, doesn't take up any screen space at all in the normal case, and can have actual words explaining what the problem is.

I agree that a dialog would be the more sensible choice here, and would be more in line with 'standard' GUI expectations. 

If so desired, the words explaining the problem could also nudge the user to check the console for additional details, with the aim of getting them into the habit of paying attention to the console, ± link to contextual help.
